decimal.go

```diff
@@ -301,6 +303,44 @@ func (d Decimal) Mod(d2 Decimal) Decimal {
 	return d.Sub(d2.Mul(quo))
 }
 
+// Sqrt returns the square root of d. The method will panic if d
+// is a negative number.
+//
+// The square root calculation is implemented using Newton's Method.
+// We start with an initial estimate for sqrt(d), and then iterate:
+//     x_{n+1} = 1/2 * ( x_n + (d / x_n) ).
+func (d Decimal) Sqrt() Decimal {
+	d.ensureInitialized()
+
+	// Validate the sign of d.
+	switch d.BigInt().Sign() {
+	case -1:
+		panic(fmt.Sprintf("square root of negative number: %v", d))
+	case 0:
+		return Zero
+	}
+
+	// Use half as the initial estimate.
+	x := d.Mul(half)
+
+	// Determine the maximum number of iterations needed.
+	scale := len(d.value.String())
+	if d.exp > 0 {
+		scale += int(d.exp)
+	}
+	steps := scale + DivisionPrecision
+
+	// Iterate.
+	for i := 0; i <= steps; i++ {
+		t := d.Div(x)   // t = d / x_n
+		t = x.Add(t)    // t = x_n + (d / x_n)
+		x = half.Mul(t) // x_{n+1} = 0.5 * t
```

Comment:
Yeah this process has made me wonder if we should reconsider using the other [decimal library](https://godoc.org/gopkg.in/inf.v0), which seems to be much more efficient in terms of memory allocations. I'm going to investigate now that I have a better idea of what specific needs we have for our decimal library.